cpp/merkletree/sparse_merkle_tree.cc

```diff
@@ -0,0 +1,214 @@
+  return ret;
+}
+
+
+Path PathFromBytes(const std::string& bytes) {
+  Path path;
+  for (int i(0); i < path.size(); ++i) {
+    path[i] = (bytes[i / 8] >> (i % 8)) & 0x01;
+  }
+  return path;
+}
+
+
+SparseMerkleTree::SparseMerkleTree(SerialHasher* hasher)
+    : serial_hasher_(CHECK_NOTNULL(hasher)->Create()),
+      treehasher_(hasher),
+      null_hashes_(CalculateNullHashes(treehasher_)),
+      root_hash_(null_hashes_[0]) {
+}
+
+
+void SparseMerkleTree::EnsureHaveLevel(size_t level) {
+  if (tree_.size() < (level + 1)) {
+    tree_.emplace_back(unordered_map<IndexType, TreeNode>());
+  }
+}
+
+
+void SparseMerkleTree::SetLeafByHash(const string& key, const string& data) {
+  serial_hasher_->Reset();
+  serial_hasher_->Update(key);
+  return SetLeaf(PathFromBytes(serial_hasher_->Final()), data);
+}
+
+
+void SparseMerkleTree::SetLeaf(const Path& path, const string& data) {
+  CHECK_EQ(treehasher_.DigestSize() * 8, path.size());
+  // Mark the tree dirty:
+  root_hash_.clear();
+
+  IndexType node_index(0);
+  for (int depth(0); depth <= kDigestSizeBits; ++depth) {
+    size_t path_index(kDigestSizeBits - depth - 1);
+    int side(path[path_index]);
+    node_index += side;
+    EnsureHaveLevel(depth);
+    auto it(tree_[depth].find(node_index));
+    if (it == tree_[depth].end()) {
+      CHECK(tree_[depth]
+                .emplace(make_pair(node_index, TreeNode(path, data)))
+                .second);
+      return;
+    } else if (it->second.type_ == TreeNode::INTERNAL) {
+      // Mark the internal node hash dirty
+      it->second.hash_.reset();
+    } else if (it->second.leaf_->path_ == path) {
+      // replacement
+      it->second.leaf_->value_ = data;
+      return;
+    } else {
+      // restructure: push the existing node down a level and replace this one
+      // with an INTERNAL node
+      EnsureHaveLevel(depth + 1);
+      IndexType child_index((node_index << 1) +
+                            it->second.leaf_->path_[path_index - 1]);
+      CHECK(tree_[depth + 1]
+                .emplace(make_pair(child_index, std::move(it->second)))
+                .second);
+      tree_[depth].erase(it);
+      CHECK(tree_[depth].emplace(make_pair(node_index, TreeNode())).second);
+    }
+    node_index <<= 1;
+  }
+  CHECK(false) << "Set failed.";
+}
+
+
+void SparseMerkleTree::DumpTree(ostream* os, size_t depth,
+                                IndexType index) const {
+  if (tree_.size() <= depth) {
+    return;
+  }
+  const string indent((depth + 1) * 2, '-');
+  for (int side(0); side < 2; ++side) {
+    auto child(tree_[depth].find(index + side));
+    if (child != tree_[depth].end()) {
+      *os << indent << side << ": " << child->second.DebugString() << "\n";
+      DumpTree(os, depth + 1, (index + side) << 1);
+    }
+  }
+}
+
+
+string SparseMerkleTree::Dump() const {
+  ostringstream ret;
+  ret << "\nTree [Root: " << util::ToBase64(root_hash_) << "]:\n";
+  if (!tree_.empty()) {
+    DumpTree(&ret, 0, 0);
```

Comment:
Yeeeaaaahhhh, it's fine :)
(note that it doesn't print the *whole* tree - just the tree representation as it is in memory, so no "default" nodes are printed, and the leaves are printed in their "unique prefix" node homes, e.g.:

```
Tree [Root: UxrRt9C5yoOtWK2JWCMLVQD5Ks923n8AftFItEdlP6Q=]:
--0: [TreeNode LEAF leaf: [Leaf path: 0101111101010101110100110101010010111001000100111001001110101011101001000011001010010010100010110011000110111010001011100110110100001100101010010100101000110110101010100010100111001000010101100100001010110010001111011100000110110000100111010101101110001011 value: three]]
--1: [TreeNode INTL hash: (unset)]
----1: [TreeNode INTL hash: (unset)]
------1: [TreeNode INTL hash: (unset)]
--------0: [TreeNode LEAF leaf: [Leaf path: 1110110100110001111101000011111111010111101011011100000001000011011100010110111000001100111010100011010000110010000100101000011110001000110011010110011011101110001110100000101100000010001111001000000010111011010000000011010110101101110000111001001001110110 value: one]]
--------1: [TreeNode LEAF leaf: [Leaf path: 1111001111100010100001011110011011011011101100000111011011010011110100111101011111000001000111001101010011101101100011010110110001100101111100000000111111110011011100011001111111011001110000001110001001110000010110000111010011111110110011001100010000111111 value: two]]
```